### PR TITLE
Comparer intermediate value

### PR DESCRIPTION
**Pull request by asi1024**
Depends on ~#400, #405,~ #406 ~and #415~

### CODE REVIEW

File: pytorch_pfn_extras/utils/comparer.py [hunk 1]
```diff
@@ -475,8 +487,15 @@ def add_engine(self, name, engine, *args, **kwargs):
         if name in self._engines.keys():
             raise ValueError(f"Engine named {name} already registered")
 
+        engine.handler = _ComparableHandler(
+            engine.handler, name, self._compare_targets, self._trigger)
+
+        child_evaluator = getattr(engine, 'evaluator', None)
+        if child_evaluator is not None:
+            # For trainer with evaluator
+            child_evaluator.handler = engine.handler
```
Comment (kmaehashi) [inline pytorch_pfn_extras/utils/comparer.py:496]: Should fail when different handler is set between trainer and evaluator.

File: pytorch_pfn_extras/utils/comparer.py [hunk 2]
```diff
@@ -497,6 +516,14 @@ def dump(self, engine, dir, train_loader, val_loader):
     def compare(self):
         """Compares outputs.
         """
+        old_backward_outputs = {
+            name: engine.handler._handler._logic.backward_outputs
+            for name, (engine, _, _) in self._engines.items()}
+        for engine, _, _ in self._engines.values():
+            logic = engine.handler._handler._logic
+            if logic.backward_outputs is None:
+                logic.backward_outputs = ['out0']
```
Comment (kmaehashi) [inline pytorch_pfn_extras/utils/comparer.py:525]: Question: What is this for?

**Review by kmaehashi — CHANGES_REQUESTED** [review]
Could you rebase? 🙏🏼

File: pytorch_pfn_extras/utils/comparer.py [hunk 3]
```diff
@@ -35,26 +41,38 @@ def train_epoch_begin(self, trainer, loader):
...
 
     def eval_loop_begin(self, evaluator):
         return self._handler.eval_loop_begin(evaluator)
 
     def eval_step(self, evaluator, batch_idx, batch, complete_fn):
-        return self._handler.eval_step(
-            evaluator, batch_idx, batch, complete_fn)
+        _thread_local.intermediate_values = {}
+        self._handler.eval_step(evaluator, batch_idx, batch, complete_fn)
+        del _thread_local.intermediate_values
```
[the remark this answers is not in the document]
Comment (asi1024) [reply pytorch_pfn_extras/utils/comparer.py]: Just to reduce memory usage.

File: pytorch_pfn_extras/utils/comparer.py (hunk 1 above)
Comment (asi1024) [reply pytorch_pfn_extras/utils/comparer.py]: This fix was already merged in https://github.com/pfnet/pytorch-pfn-extras/pull/425. I will fix it in another PR.

**Comment by asi1024** [comment]
/test
@kmaehashi PTAL!

File: pytorch_pfn_extras/utils/comparer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kmaehashi) [reply pytorch_pfn_extras/utils/comparer.py]: Ah, ok now understood after reading the updated `intermediate_value` API :smile:

File: pytorch_pfn_extras/utils/comparer.py (hunk 1 above)
Comment (kmaehashi) [reply pytorch_pfn_extras/utils/comparer.py]: For record: #432

**Review by kmaehashi — APPROVED** [review]

**Comment by asi1024** [comment]
Close https://github.com/pfnet/pytorch-pfn-extras/issues/353.

**Pull request merged**